scipy/cluster/hierarchy.py

```diff
@@ -1395,6 +1400,208 @@ def cut_tree(Z, n_clusters=None, height=None):
     return groups.T
 
 
+def cut_tree_balanced(Z, max_cluster_size):
+    """
+    Given a linkage matrix Z and max cluster size, return a balanced cut tree.
+
+    The function looks recursively along the hierarchical tree, from the root
+    (single cluster gathering all the samples) to the leaves (i.e. the clusters
+    with only one sample), retrieving the biggest possible clusters containing
+    a number of samples lower than a given maximum. If a cluster at a specific
+    tree level contains a number of samples higher than the given maximum, it
+    is ignored and its offspring (smaller) sub-clusters are taken into
+    consideration. If the cluster contains a number of samples lower than the
+    given maximum, it is taken as result and its offspring sub-clusters not
+    further processed.
+
+    Parameters
+    ----------
+    Z : ndarray
+        The linkage matrix resulting from calling `ward` or `linkage`. I.e. it
+        contains the hierarchical clustering encoded as a linkage matrix.
+    max_cluster_size : int
+        Maximum number of data samples contained within the resulting clusters.
+        Thus, all resulting clusters will contain a number of data samples
+        ``<= max_cluster_size``. Must be >= 1.
+
+    Returns
+    -------
+    cluster_id : ndarray
+        One-dimensional array of integers containing for each input sample its
+        corresponding cluster id. The cluster id is an integer which is higher
+        for deeper tree levels.
+    cluster_level : ndarray
+        One-dimensional array of integer arrays containing for each input
+        sample its corresponding cluster tree level, i.e. a sequence of
+        0's and 1's. Note that the cluster level is longer for deeper tree
+        levels, being [0] the root cluster, [0, 0] and [0, 1] its offspring,
+        and so on. Also note that in each cluster splitting, the label 0
+        denotes the bigger cluster, while the label 1 denotes the smallest.
+
+    See Also
+    --------
+    cut_tree
+
+    Notes
+    -----
+    There are several implemented methods following the same idea, i.e.
+    performing a tree cut in which the resulting clusters are at different tree
+    levels, but using more elaborated algorithms (in which the threshold of
+    max_cluster_size is dinamycally computed). The CRAN R package
+    [dynamicTreeCut](https://github.com/cran/dynamicTreeCut) implements novel
+    dynamic branch cutting methods for detecting clusters in a dendrogram
+    depending on their shape. Further, [MLCut]](https://github.com/than8/MLCut)
+    provides interactive methods to cut tree branches at multiple levels.
+    Note that in the present method, the max_cluster_size threshold is a fixed
+    value given as input.
+
+    Further, note that this algorithm uses :math:`O(n^2)` memory, i.e. the same
+    as cut_tree because a full cut tree of the linkage matrix is performed
```

Comment:
single backticks around `cut_tree`

Reply:
Fixed (will be pushed when the current CI job is over).